java/ql/lib/semmle/code/java/regex/RegexTreeView.qll

```diff
@@ -0,0 +1,1031 @@
+  /** A quoted sequence */
+  TRegExpQuote(Regex re, int start, int end) { re.quote(start, end) } or
+  /** A back reference */
+  TRegExpBackRef(Regex re, int start, int end) { re.backreference(start, end) }
+
+/**
+ * An element containing a regular expression term, that is, either
+ * a string literal (parsed as a regular expression)
+ * or another regular expression term.
+ */
+class RegExpParent extends TRegExpParent {
+  /** Gets a textual representation of this element. */
+  string toString() { result = "RegExpParent" }
+
+  /** Gets the `i`th child term. */
+  abstract RegExpTerm getChild(int i);
+
+  /** Gets a child term . */
+  RegExpTerm getAChild() { result = this.getChild(_) }
+
+  /** Gets the number of child terms. */
+  int getNumChild() { result = count(this.getAChild()) }
+
+  /** Gets the associated regex. */
+  abstract Regex getRegex();
+}
+
+/** A string literal used as a regular expression */
+class RegExpLiteral extends TRegExpLiteral, RegExpParent {
+  Regex re;
+
+  RegExpLiteral() { this = TRegExpLiteral(re) }
+
+  override RegExpTerm getChild(int i) { i = 0 and result.getRegex() = re and result.isRootTerm() }
+
+  /** Holds if dot, `.`, matches all characters, including newlines. */
+  predicate isDotAll() { re.getAMode() = "DOTALL" }
+
+  /** Holds if this regex matching is case-insensitive for this regex. */
+  predicate isIgnoreCase() { re.getAMode() = "IGNORECASE" }
+
+  /** Get a string representing all modes for this regex. */
+  string getFlags() { result = concat(string mode | mode = re.getAMode() | mode, " | ") }
+
+  override Regex getRegex() { result = re }
+
+  /** Gets the primary QL class for this regex. */
+  string getPrimaryQLClass() { result = "RegExpLiteral" }
+}
+
+/**
+ * A regular expression term, that is, a syntactic part of a regular expression.
+ */
+class RegExpTerm extends RegExpParent {
+  Regex re;
+  int start;
+  int end;
+
+  RegExpTerm() {
+    this = TRegExpAlt(re, start, end)
+    or
+    this = TRegExpBackRef(re, start, end)
+    or
+    this = TRegExpCharacterClass(re, start, end)
+    or
+    this = TRegExpCharacterRange(re, start, end)
+    or
+    this = TRegExpNormalChar(re, start, end)
+    or
+    this = TRegExpQuote(re, start, end)
+    or
+    this = TRegExpGroup(re, start, end)
+    or
+    this = TRegExpQuantifier(re, start, end)
+    or
+    this = TRegExpSequence(re, start, end)
+    or
+    this = TRegExpSpecialChar(re, start, end)
+  }
+
+  /**
+   * Gets the outermost term of this regular expression.
+   */
+  RegExpTerm getRootTerm() {
+    this.isRootTerm() and result = this
+    or
+    result = this.getParent().(RegExpTerm).getRootTerm()
+  }
+
+  /**
+   * Holds if this term is part of a string literal
+   * that is interpreted as a regular expression.
+   */
+  predicate isUsedAsRegExp() { any() }
+
+  /**
+   * Holds if this is the root term of a regular expression.
+   */
+  predicate isRootTerm() { start = 0 and end = re.getText().length() }
```

Comment:
This indicates to me that `start` is inclusive and `end` is exclusive`. Presumably that's the convention that's followed for all these start-end-ranges? Could we document this?

Reply:
Nevermind, found the comment at the top of `regex.qll`.